h/services/job_queue/queue.py

```diff
@@ -0,0 +1,154 @@
+            application-level URL-safe format
+        """
+        where = [Annotation.id == annotation_id]
+        self.add_where(where, tag, Priority.SINGLE_ITEM, force, schedule_in)
+
+    def add_by_user(self, userid, tag, force=False, schedule_in=None):
+        """
+        Queue all a user's annotations.
+
+        See Queue.add() for documentation of the params.
+
+        :param userid: The ID of the user in "acct:USERNAME@AUTHORITY" format
+        :type userid: unicode
+        """
+        where = [Annotation.userid == userid]
+        self.add_where(where, tag, Priority.SINGLE_USER, force, schedule_in)
+
+    def add_by_group(self, groupid, tag, force=False, schedule_in=None):
+        """
+        Queue all annotations in a group.
+
+        See Queue.add() for documentation of the params.
+
+        :param groupid: The pubid of the group
+        :type groupid: unicode
+        """
+        where = [Annotation.groupid == groupid]
+        self.add_where(where, tag, Priority.SINGLE_GROUP, force, schedule_in)
+
+    def add_where(  # pylint: disable=too-many-arguments
+        self,
+        where,
+        tag,
+        priority,
+        force=False,
+        schedule_in=None,
+        name="sync_annotation",
+    ):
+        """
+        Queue annotations matching a filter .
+
+        :param where: A list of SQLAlchemy BinaryExpression objects to limit
+            the annotations to be added
+        :param tag: The tag to add to the job on the queue. For documentation
+            purposes only
+        :param priority: Integer priority value (higher number is lower
+            priority)
+        :param force: Whether to force reindexing of the annotation even if
+            it's already indexed
+        :param schedule_in: A number of seconds from now to wait before making
+            the job available for processing. The annotation won't be synced
+            until at least `schedule_in` seconds from now
+        :param name : Name of the task in the queue
+        """
+        where_clause = and_(*where) if len(where) > 1 else where[0]
+        schedule_at = datetime.utcnow() + timedelta(seconds=schedule_in or 0)
+
+        query = Job.__table__.insert().from_select(
+            [Job.name, Job.scheduled_at, Job.priority, Job.tag, Job.kwargs],
+            select(
+                [
+                    literal_column(f"'{name}'"),
+                    literal_column(f"'{schedule_at}'"),
+                    literal_column(str(priority)),
+                    literal_column(repr(tag)),
+                    func.jsonb_build_object(
+                        "annotation_id", Annotation.id, "force", bool(force)
+                    ),
+                ]
+            ).where(where_clause),
+        )
+
+        self._db.execute(query)
+        mark_changed(self._db)
+
+    @staticmethod
+    def queue_annotations_between_times(start_time, end_time, tag):
```

Comment:
These don't add much but are easier to mock (instead of patching) via the mock service when used elsewhere.

Reply:
Gone for the route of calling the tasks directly.

A couple of services were also calling these but it's not a big deal. 

It's not the goal metric but this change reduced line count.